### PR TITLE
contrib.seq2seq initial commit

### PR DESCRIPTION
**Pull request by alrojo**
As advised by @lukaszkaiser from #4686, initial commit of the `tf.contrib.seq2seq` folder.

The purpose of this pull request is to agree on a skeleton for the `tf.contrib.seq2seq` folder, I took inspiration from the [`tf.contrib.layers`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/) folder and the [`tf.contrib.losses`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/losses/) folder.

After this skeleton has been accepted, I will start by filling out the `tf.contrib.seq2seq.loss` with #4382 , the `tf.contrib.seq2seq.rnn_decoder` with #4686 and a `tf.contrib.seq2seq.rnn_decoder_attention` including `kernel_test.loss_ops_test.py` and `kernel_test.layers_ops_test.py`.

Please note, as some of the calls (BUILD?) for the contrib section are scattered across various folders, I might be missing some calls or definitions somewhere.

### CODE REVIEW

**Comment by lukaszkaiser** [comment]
This looks good to me, great thanks! Let's test it and get it in.

**Comment by alrojo** [comment]
@lukaszkaiser Would that require any action from me?

**Comment by lukaszkaiser** [comment]
I think not. Can one of the admins test and merge this please?

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by alrojo** [comment]
Aha! I forgot to add the `all_files` symlink in the `contrib/seq2seq/BUILD` file. I just pushed it, sorry for the inconvenience! Please try and rerun @jhseu 

On a related note: Do you have a guide on how to run these "Jenkins" test locally, so I can skip the trouble of having your admins debug for me? Thanks.

**Comment by jhseu** [comment]
You can reproduce the tests we run with tools/ci_build/ci_sanity.sh tools/ci_build/ci_parameterized_build.sh. Jenkins just runs those with different environment variables set. We definitely should document it better for contributors, though.

Jenkins, test this please

**Comment by yifeif** [comment]
@tensorflow-jenkins test it please.

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please.

**Comment by alrojo** [comment]
I have switched the style of coding to `tf.contrib.bayesflow` instead, as it seemed more matching (there is quite a few styles in the contrib folder).

I couldn't get the tests @jhseu pointed out to work. But I ran the testing guide (step 1-4) from: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/README.md

It was very time consuming (~6 hours per run) but my last commit had no FAILED.

@jhseu I would be interested in making a short tutorial for how to setup the testing environment (e.g. on a fresh ubuntu 14.04 GCP/AWS instance). Could you perhaps assist me in the steps required to make a complete replica of the Jenkins test environment? (I can make a separate issue for this if wanted)

Fresh Ubuntu 14.04

- <install Docker>
- git clone https://github.com/tensorflow/tensorflow.git
- cd tensorflow
- ??

**Comment by ethancaballero** [comment]
@alrojo Does your `dynamic_rnn_decoder_attention` do beam-search in the graph?
Also, I can help out with adding a multi-hop attention option (similar to Memory Networks) to `seq2seq` for use cases that require transitive reasoning about inputs.

**Comment by alrojo** [comment]
@ethancaballero The `dynamic_rnn_decoder_attention` will not have beam search implented (yet!). Multi-layered decoder and beam search is first on my todo list.

When this PR has been merged I will update the two other PRs (`sequence_loss` and `rnn_decoder`) and make a new with the `dynamic_rnn_decoder_attention` where we can discuss extensions. By multi-hop attention are you referring to: $m_i$ in (https://arxiv.org/abs/1506.07285)?

**Review by lukaszkaiser — APPROVED** [review]
Looks good to me, thanks!

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please.

**Comment by ethancaballero** [comment]
@alrojo Yes, although I would instead use the Attention based GRU from equation 11 of https://arxiv.org/pdf/1603.01417.pdf to obtain contextual vector c_t that then updates episode memory m_t−1 (via a vanilla GRU) to produce m_t. There's an example tensorflow implementation of Attention based GRU here: 
https://github.com/therne/dmn-tensorflow/blob/master/utils/attn_gru.py

Alternatively, one could use a weighted sum as a soft attention to compute contextual vector c_t during each hop (which is more similar to the current tf.nn.seq2seq api), but this weighted sum method loses positional & ordering information and as a result is significantly less accurate for use cases that require more than 2 hops.

**Comment by alrojo** [comment]
Okay, that the `Sanity Checks` failed because the filegroups wasnt ordered alphabetically (I think). I have now updated the filegroups so the `seq2seq` is alphabetically correctly placed in the BUILD's filegroup.

This was the `FAILED` stack trace from the `Sanity Checks`:
```
=== Sanity check step 3 of 4: do_buildifier (buildifier check) ===

Running do_buildifier on 112 files

tensorflow/BUILD # listsort unsafesort sort:filegroup.data

buildifier took 1 s

FAIL: buildifier found errors and/or warnings in above BUILD files.
buildifier suggested the following changes:
123a124
>         "//tensorflow/contrib/seq2seq:all_files",
126d126
<         "//tensorflow/contrib/seq2seq:all_files",
Please fix manually or run buildifier <file> to auto-fix.
```

**Comment by alrojo** [comment]
@yifeif could I have you try test it again? :) thanks

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by lukaszkaiser** [comment]
Please merge this when possible guys!
